### PR TITLE
Showimage

### PR DESCRIPTION
**Pull request by esteve**
Adds an executable to visualize images coming from a subscription.

Depends on #2 

Connects to ros2/ros2#54

### CODE REVIEW

**Comment by esteve** [comment]
@dirk-thomas @jacquelinekay @tfoote @wjwwood ping

File: cam2image/src/cam2image.cpp [hunk 1]
```diff
@@ -99,6 +99,7 @@ int main(int argc, char * argv[])
         convert_frame_to_message(flipped_frame, i, msg);
       }
       std::cout << "Publishing image #" << i << std::endl;
+      msg->header.frame_id = std::to_string(i);
```
Comment (dirk-thomas) [inline cam2image/src/cam2image.cpp]: Why is this added here? That is already done in https://github.com/ros2/demos/blob/4f409e9e89796143a94b9fa1cd18988d64edb64f/cam2image/src/cam2image.cpp#L52

File: cam2image/src/showimage.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  } else if(encoding == "rgb8") {
+    return CV_8UC3;
+  } else if(encoding == "mono16") {
+    return CV_16SC1;
+  } else if(encoding == "rgba8") {
+    return CV_8UC4;
+  } else {
+    throw std::runtime_error("Unsupported encoding type");
+  }
+}
```
Comment (dirk-thomas) [inline cam2image/src/showimage.cpp]: The same code already exists in the other cpp file. I would suggest moving it to a header and reusing it in both.

File: cam2image/src/showimage.cpp [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+void showImage(const sensor_interfaces::msg::Image::ConstSharedPtr & msg) {
+  std::cout << "Received image #" << msg->header.frame_id << std::endl;
```
Comment (dirk-thomas) [inline cam2image/src/showimage.cpp]: `printf` instead.

File: cam2image/src/showimage.cpp (hunk 2 above)
Comment (dirk-thomas) [inline cam2image/src/showimage.cpp]: Never mind - this is the opposite direction.

File: cam2image/src/showimage.cpp [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+  cv::imshow("view", frame);
+  cv::waitKey(1);
+}
+
+int main(int argc, char * argv[])
+{
+  rclcpp::init(argc, argv);
+
+  auto node = rclcpp::node::Node::make_shared("cam2image");
```
Comment (dirk-thomas) [inline cam2image/src/showimage.cpp]: This should have a different node name.

File: cam2image/src/showimage.cpp [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+  } else if(encoding == "mono16") {
+    return CV_16SC1;
+  } else if(encoding == "rgba8") {
+    return CV_8UC4;
+  } else {
+    throw std::runtime_error("Unsupported encoding type");
+  }
+}
+
+void showImage(const sensor_interfaces::msg::Image::ConstSharedPtr & msg) {
```
Comment (dirk-thomas) [inline cam2image/src/showimage.cpp]: Should this be lower case underscore instead?

**Comment by dirk-thomas** [comment]
The node for showing images is a bit weird in the `cam2image` package. Should we either split it into a separate packages or rename this package to describe better what it will contain?

**Comment by esteve** [comment]
@dirk-thomas I'd rename the package, what would be a suitable name? `imagetools` ?

**Comment by dirk-thomas** [comment]
`image_tools` sounds fine to me.

File: cam2image/src/cam2image.cpp (hunk 1 above)
Comment (esteve) [inline cam2image/src/cam2image.cpp]: Removed.

File: cam2image/src/showimage.cpp (hunk 3 above)
Comment (esteve) [inline cam2image/src/showimage.cpp]: Used a `cout` and a `stringstream` to avoid interleaving.

File: cam2image/src/showimage.cpp (hunk 4 above)
Comment (esteve) [inline cam2image/src/showimage.cpp]: Fixed.

File: cam2image/src/showimage.cpp (hunk 5 above)
Comment (esteve) [inline cam2image/src/showimage.cpp]: Done.

**Comment by esteve** [comment]
@dirk-thomas I just renamed the package to `image_tools`

File: cam2image/src/showimage.cpp (hunk 3 above)
Comment (dirk-thomas) [inline cam2image/src/showimage.cpp]: This could be a single line with `printf` instead of three.

File: image_tools/package.xml [hunk 6]
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0"?>
 <package format="2">
-  <name>cam2image</name>
+  <name>image_tools</name>
   <version>0.0.0</version>
   <description>A node to read images form a camera and publish them on a topic.</description>
```
Comment (tfoote) [inline image_tools/package.xml]: The description could use a slight extension for the new scope.

**Comment by tfoote** [comment]
One comment on the description but otherwise lgtm

File: image_tools/package.xml (hunk 6 above)
Comment (esteve) [inline image_tools/package.xml]: Done.

**Comment by wjwwood** [comment]
lgtm

File: image_tools/package.xml (hunk 6 above)
Comment (dirk-thomas) [inline image_tools/package.xml]: This comments does not seem to be addressed in the merged patch yet.

Comment (esteve) [inline image_tools/package.xml]: I committed it to the qos branch by mistake.

Comment (esteve) [inline image_tools/package.xml]: Fixed in master.

File: image_tools/src/showimage.cpp [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    return CV_16SC1;
+  } else if(encoding == "rgba8") {
+    return CV_8UC4;
+  } else {
+    throw std::runtime_error("Unsupported encoding type");
+  }
+}
+
+void show_image(const sensor_interfaces::msg::Image::ConstSharedPtr & msg) {
+  std::stringsteam ss;
```
Comment (dirk-thomas) [inline image_tools/src/showimage.cpp]: This change breaks the build. Please fix the syntax.

Comment (jacquelinekay) [inline image_tools/src/showimage.cpp]: I committed the fix for this syntax error to master in b603bfe739e2021ec7b64a0f34c349e02a57377d.

Comment (esteve) [inline image_tools/src/showimage.cpp]: @jacquelinekay thanks! I pushed a change to include `#include <sstream>` right at the same time.

**Comment by dirk-thomas** [comment]
@esteve With the windows build being back up again please revisit the new warnings introduced by this PR: http://ci.ros2.org/job/ros2_batch_ci_windows/133/warnings34Result/category.63481637/package.-2136935613/

**Comment by esteve** [comment]
@dirk-thomas I committed this, but haven't merged it to master:

https://github.com/ros2/demos/commit/f3ab25dbe4cd928d1b0073442a7bbf67f2e9f82a

this removes the warning you linked to, though it doesn't fix the existing warnings.

**Comment by dirk-thomas** [comment]
Looking at the surrounding code it looks like that the allocated `data` is leaked for every frame. Before continuing on the iterator the first thing should be to fix the memory leak. May be loaning the ROS image data directly to OpenCV will avoid the whole copy?

**Comment by esteve** [comment]
@dirk-thomas I changed it to use the `data()` pointer from the `std::vector`:

https://github.com/ros2/demos/commit/38e89cd6b4c8e01284d33ba4eea11c66b8138721

I'll start a Windows CI job when Jose is done with the Windows slave.
